pkg/bootstrap/bootstrap.go

```diff
@@ -0,0 +1,176 @@
+package bootstrap
+
+import (
+	"context"
+
+	"github.com/solo-io/skv2/pkg/stats"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/spf13/pflag"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/go-logr/zapr"
+	"github.com/solo-io/go-utils/contextutils"
+	v1 "github.com/solo-io/skv2/pkg/api/core.skv2.solo.io/v1"
+	"github.com/solo-io/skv2/pkg/multicluster"
+	"github.com/solo-io/skv2/pkg/multicluster/watch"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	zaputil "sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	// required import to enable kube client-go auth plugins
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+)
+
+// StartParameters specify paramters for starting a generic controller which may need access to its local cluster as well as remote (multicluster) clients and managers
+type StartParameters struct {
+	Ctx             context.Context
+	MasterManager   manager.Manager
+	McClient        multicluster.Client    // nil if running in agent mode
+	Clusters        multicluster.Interface // nil if running in agent mode
+	SnapshotHistory *stats.SnapshotHistory
+	// Reference to Settings object this controller uses.
+	SettingsRef v1.ObjectRef
+	// enable additional logging
+	VerboseMode bool
+}
+
+// the start function that will be called with the initialized parameters
+type StartFunc func(
+	parameters StartParameters,
+) error
+
+// bootstrap options for starting discovery
+type Options struct {
+	// MetricsBindPort is the TCP port that the controller should bind to
+	// for serving prometheus metrics.
+	// It can be set to 0 to disable the metrics serving.
+	MetricsBindPort uint32
+
+	// MasterNamespace if specified restricts the Master manager's cache to watch objects in the desired namespace.
+	// Defaults to all namespaces.
+	//
+	// Note: If a namespace is specified, controllers can still Watch for a cluster-scoped resource (e.g Node).  For namespaced resources the cache will only hold objects from the desired namespace.
```

Comment:
We should move away from master. maybe `MgmtNamespace`

Reply:
it's a bit confusing with other uses of management/manager in the code. i don't think there's a clear path forward here, and since this code is already written (just moving repos to make it shared) i'd prefer not to change